ubuntu_image/tests/test_parser.py

```diff
@@ -526,6 +496,151 @@ def test_volume_filesystem_default_none(self):
         partition0 = volume0.structures[0]
         self.assertEqual(partition0.filesystem, FileSystemType.none)
 
+    def test_volume_filesystem_bad(self):
+        with ExitStack() as resources:
+            cm = resources.enter_context(
+                self.assertRaises(GadgetSpecificationError))
+            parse("""\
+volumes:
+  first-image:
+    bootloader: u-boot
+    structure:
+        - type: 00000000-0000-0000-0000-0000deadbeef
+          size: 400M
+          filesystem: zfs
+""")
+        self.assertEqual(
+            str(cm.exception),
+            ("Invalid gadget.yaml value 'zfs' @ "
+             'volumes:<volume name>:structure:<N>:filesystem'))
+
+    def test_volume_structure_role(self):
+        gadget_spec = parse("""\
+volumes:
+  first-image:
+    schema: mbr
+    bootloader: u-boot
+    structure:
+        - type: ef
+          size: 100
+          role: mbr
+  second-image:
+    structure:
+        - type: 00000000-0000-0000-0000-0000feedface
+          size: 200
+          role: system-boot
+  third-image:
+    structure:
+        - type: 00000000-0000-0000-0000-0000deafbead
+          size: 300
+          role: system-data
+  fourth-image:
+    structure:
+        - type: 00000000-0000-0000-0000-0000deafbead
+          size: 400
+""")
+        self.assertEqual(len(gadget_spec.volumes), 4)
+        self.assertEqual({
+            'first-image': StructureRole.mbr,
+            'second-image': StructureRole.system_boot,
+            'third-image': StructureRole.system_data,
+            'fourth-image': None,
+            },
+            {key: gadget_spec.volumes[key].structures[0].role
+             for key in gadget_spec.volumes}
+            )
+
+    def test_volume_structure_invalid_role(self):
+        with ExitStack() as resources:
+            cm = resources.enter_context(
+                self.assertRaises(GadgetSpecificationError))
+            parse("""\
+volumes:
+  first-image:
+    bootloader: u-boot
+    structure:
+        - type: 00000000-0000-0000-0000-0000deadbeef
+          size: 100
+          role: foobar
+""")
+        self.assertEqual(
+            str(cm.exception),
+            ("Invalid gadget.yaml value 'foobar' @ "
+             'volumes:<volume name>:structure:<N>:role'))
+
+    def test_volume_structure_mbr_role_too_big(self):
+        with ExitStack() as resources:
+            cm = resources.enter_context(
+                self.assertRaises(GadgetSpecificationError))
+            parse("""\
+volumes:
+  first-image:
+    schema: mbr
+    bootloader: u-boot
+    structure:
+        - type: ef
+          size: 1M
```

Comment:
Can you add some corner case tests, e.g. with size == (445, 446, 447) where the first two are allowed, but the last one is not?